ultimarc/devices/ultrastik.py

```diff
@@ -0,0 +1,177 @@
+
+        data = UltraStikStruct()
+        data.keepAnalog = 0x11 if config.keepAnalog else 0x11
+        data.mapSize = config.mapSize
+        data.restrictor = 0x09 if config.restrictor else 0x10
+        data.borders = (ct.c_uint8 * 8)(*config.borders)
+        data.map = (ct.c_uint8 * 81)(*[DIRECTION_MAP[v] for v in config.map])
+        data.flash = 0x00 if config.flash else 0xFF
+        data.reserved = (ct.c_uint8 * 3)(*[0, 0, 0])
+
+        resp_2 = resp_3 = resp_4 = False
+        payload = (ct.c_uint8 * USTIK_PRE_MESG_LENGTH)(0)
+
+        resp_1 = self.write_raw(USBRequestCode.ULTRASTIK_E9, 0x1, USTIK_PRE_INTERFACE, ct.c_void_p(), 0x0,
+                                request_type=USBRequestType.REQUEST_TYPE_VENDOR,
+                                recipient=USBRequestRecipient.RECIPIENT_OTHER)
+        # We need to write 32 bytes here.
+        for x in range(3):
+            ct.memmove(ct.addressof(payload), ct.byref(data, USTIK_PRE_MESG_LENGTH * x), USTIK_PRE_MESG_LENGTH)
+            _logger.debug(f"  config data block {x+1}: {' '.join('%02X' % b for b in payload)}")
+
+            resp_2 = self.write(USBRequestCode.ULTRASTIK_EB, 0x0, USTIK_PRE_INTERFACE,
+                                payload, USTIK_PRE_MESG_LENGTH,
+                                request_type=USBRequestType.REQUEST_TYPE_VENDOR,
+                                recipient=USBRequestRecipient.RECIPIENT_OTHER)
+            resp_3 = self.read(USBRequestCode.ULTRASTIK_EA, 0x0, USTIK_PRE_INTERFACE, ct.c_void_p(), 0x0,
+                                request_type=USBRequestType.REQUEST_TYPE_VENDOR,
+                                recipient=USBRequestRecipient.RECIPIENT_OTHER)
+            if not resp_2 or not resp_3:
+                break
+
+        if resp_2 and resp_3:
+            resp_4 = self.write(USBRequestCode.ULTRASTIK_E9, 0x0, USTIK_PRE_INTERFACE, ct.c_void_p(), 0x0,
+                            request_type=USBRequestType.REQUEST_TYPE_VENDOR,
+                            recipient=USBRequestRecipient.RECIPIENT_OTHER)
+
+        return not False in [resp_1, resp_2, resp_3, resp_4]
+
+
+class UltraStikDevice(USBDeviceHandle):
+    """
+    Manage an UltraStik 360 Joystick (2015 or newer)
+    """
+    class_id = 'ultrastik'  # Used to match/filter devices.
+    class_descr = _('UltraStik 360 Joystick')
+    interface = USTIK_INTERFACE
+
+    def set_controller_id(self, new_controller_id: int) -> bool:
+        """
+        Set the Ultrastik controller ID
+        :param new_controller_id: New control ID to set, must be an integer between 1 and 4.
+        :return: True if successful, otherwise false.
+        """
+        # Calculate new device ID.
+        new_id = USTIK_CONFIG_BASE + (new_controller_id - 1)
+        # Setup an array with zero for all byte values.
+        data = (ct.c_ubyte * USTIK_MESG_LENGTH)(0)
+        data[0] = new_id
+
+        resp = self.write(USBRequestCode.SET_CONFIGURATION, 0x0, USTIK_INTERFACE, data, USTIK_MESG_LENGTH,
```

Comment:
0x0 needs to be 0x0200

Reply:
I think we messed up with the 'self.write() method', it always auto adds 0x0200 to every call for the 'report_id' arg.  I think we assumed that every write call would require a 0x0200 value for the 'w_value' argument value to 'self._make_control_transfer()' (See: _device.py:266), and we hard coded it into the 'self.write()' method. This is why I ended up adding the 'self.write_raw()' method to not force adding 0x0200 to the 'report_id' arg.  When reprogramming the ultrastik device ID, we need to make a write call without a hardcoded '0x0200' value in the 'report_id'.

Reply:
The 2015 and newer devices do use the 0x200 in w_value.  I had to make that change to get the 2015 ultrastik to work, the id change and configuration change.  The original assumption was to support the 2015 and newer devices and let the c++ code continue to support the older devices.  Also since we started with ipac devices that also push us into this decision.  That being said I don't mind supporting the previous generation in this code.  We would have to rework/reorganize the write methods.